### PR TITLE
Refactoring: Strict TypeScript Fix #1208

### LINKED ISSUE #1208: [Discussion] Enable stricter TypeScript checks
**Is your feature request related to a problem? Please describe.**

There are currently no additional TypeScript checks enabled in .tsconfig (https://github.com/REditorSupport/vscode-R/blob/master/tsconfig.json).

Using strict-mode in TypeScript is recommended: https://www.typescriptlang.org/tsconfig#strict

Opting out on a per-case basis via `// @ts-ignore` or `foo as FooType` should be preferred to indicate to other developers that a statement is not an error but a conscious decision.

**Describe the solution you'd like**

The VSCode Python extension ([see](https://github.com/microsoft/vscode-python/blob/main/tsconfig.json)) uses the following settings:

```js
    "compilerOptions": {
        "strict": true,
        "noImplicitAny": true,
        "noImplicitThis": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noFallthroughCasesInSwitch": true,
    },
```

Enabling the options below shows quite a number of errors:

```sh
> yarn run compile
Found 302 errors in 32 files.

Errors  Files
    16  src/completions.ts:35
    16  src/extension.ts:35
     2  src/helpViewer/cran.ts:83
     2  src/helpViewer/helpProvider.ts:301
     6  src/helpViewer/index.ts:175
     3  src/helpViewer/packages.ts:85
    11  src/helpViewer/panel.ts:72
    15  src/helpViewer/treeView.ts:68
     6  src/languageService.ts:42
     2  src/lineCache.ts:30
     3  src/lintrConfig.ts:12
     5  src/liveShare/index.ts:21
     2  src/liveShare/shareCommands.ts:80
    11  src/liveShare/shareSession.ts:68
    17  src/liveShare/shareTree.ts:14
     5  src/plotViewer/index.ts:289
     3  src/preview.ts:32
     1  src/rGitignore.ts:37
     7  src/rmarkdown/draft.ts:49
    32  src/rmarkdown/index.ts:63
    13  src/rmarkdown/knit.ts:12
    11  src/rmarkdown/manager.ts:32
    23  src/rmarkdown/preview.ts:20
     5  src/rstudioapi.ts:119
    26  src/rTerminal.ts:20
    12  src/selection.ts:8
    14  src/session.ts:185
     1  src/tasks.ts:104
     3  src/test/suite/index.ts:5
     2  src/test/suite/syntax.test.ts:56
    20  src/util.ts:25
     7  src/workspaceViewer.ts:39
```

(For reference, if only `"strict"` is enabled it shows `Found 295 errors in 31 files.`)

**PR**

If there is agreement on enabling strict mode I volunteer to work on a PR that fixes the existing errors. (I don't want to start working on it if its not desired)

### PR DESCRIPTION
**Pull request by nx10**
# What problem did you solve?

WIP PR to refactor the entire TypeScript codebase to use stricter type checks. See https://github.com/REditorSupport/vscode-R/issues/1208

We should work bottom-up and fix files with many dependents earlier as null-checks may propagate.

## TODO

Rough checklist to track progress:

- [x] html\help\script.ts
- [x] html\help\webviewMessages.d.ts
- [x] html\httpgd\index.ts
- [x] html\httpgd\webviewMessages.d.ts
- [x] src\api.d.ts
- [x] src\apiImplementation.ts
- [x] src\completions.ts
- [x] src\cppProperties.ts
- [x] src\extension.ts
- [x] src\languageService.ts
- [x] src\lineCache.ts
- [x] src\lintrConfig.ts
- [x] src\preview.ts
- [x] src\rGitignore.ts
- [x] src\rstudioapi.ts
- [x] src\rTerminal.ts
- [x] src\selection.ts
- [x] src\session.ts
- [x] src\tasks.ts
- [x] src\util.ts
- [x] src\workspaceViewer.ts
- [x] src\helpViewer\cran.ts
- [x] src\helpViewer\helpProvider.ts
- [x] src\helpViewer\index.ts
- [x] src\helpViewer\packages.ts
- [x] src\helpViewer\panel.ts
- [x] src\helpViewer\treeView.ts
- [x] src\helpViewer\webviewMessages.d.ts
- [x] src\liveShare\index.ts
- [x] src\liveShare\shareCommands.ts
- [x] src\liveShare\shareSession.ts
- [x] src\liveShare\shareTree.ts
- [x] src\liveShare\virtualDocs.ts
- [x] src\plotViewer\httpgdTypes.d.ts
- [x] src\plotViewer\index.ts
- [x] src\plotViewer\webviewMessages.d.ts
- [x] src\rmarkdown\draft.ts
- [x] src\rmarkdown\index.ts
- [x] src\rmarkdown\knit.ts
- [x] src\rmarkdown\manager.ts
- [x] src\rmarkdown\preview.ts
- [x] src\test\runTest.ts
- [x] src\test\suite\extendSelection.test.ts
- [x] src\test\suite\index.ts
- [x] src\test\suite\lineCache.test.ts
- [x] src\test\suite\syntax.test.ts

### CODE REVIEW

**Comment by ManuelHentschel** [comment]
I'm quite familiar with the help and plot viewer related code, so I could do those files. Should I just push the corresponding changes directly to this branch?

**Comment by nx10** [comment]
> I'm quite familiar with the help and plot viewer related code, so I could do those files. Should I just push the corresponding changes directly to this branch?

Yes, go for it. I am done for today and will continue tomorrow. I have just finished all code files in the src/ root but ` src\session.ts` (I don't really understand the request file mechanism used there so help is appreciated).

**Comment by nx10** [comment]
For reference: There was/is a pretty large number of unchecked null-ish values. (for example when the R-path is not set) I try to return early or abort gracefully where possible but this PR will need extensive testing of all vscode-r functionality after it is finished. On the plus side this will hopefully fix a lot of hard to track down bugs presently and in the future.

**Comment by nx10** [comment]
We should probably refactor all occurrences of 

`config().get<T>(section: string): T | undefined`

to

`config().get<T>(section: string, defaultValue: T): T`

**Comment by nx10** [comment]
It's done! We should definitely test this well before merge, but I will open the PR for review now.

File: src/util.ts [hunk 1]
```diff
@@ -518,3 +533,57 @@ export async function promptToInstallRPackage(name: string, section: string, cwd
...
+    if(!s){
+        return fallback;
+    }
+    if(typeof s !== 'string'){
+        // s is already ViewColumn:
+        return s;
+    }
+    if(Object.keys(vscode.ViewColumn).includes(s)){
+        const sKey = s as keyof typeof vscode.ViewColumn;
+        return(vscode.ViewColumn[sKey]);
```
Comment (nx10) [inline src/util.ts:585]: This is not entirely correct. `Object.keys(SomeEnumType)` will also include it's stringified values.
([TS Playground](https://www.typescriptlang.org/play?#code/FAUwdgrgtgBAagSxAdwMIHsA20wwN7AxHFED0AVOYSSeTAIIzkDOAnlAEZYIDGdIAEwQAXdACcYPLDhhiQABznNwwhGADmMYQAsQkiGLlhhmVjACGPVQDc9U7FDAA6GABVtCZjGvnsIajREdDzmuBx6EMoCMMi6uOjy4GqagiLizAA0MBwQwlq6TEpYtgJ0eAACmGoA1m4gAB7CAKJComJO1kho0o7eXRgOYAC+ALQ+fgGBdOgAZjCpbV7ICJiYFpjI5qxe4TAABgDyYCB7WXuuyOin++5yJxlOjzDi+xBgAiAzaoJ72bkwx1sEj29CsCFseyckxo5FI0OIoJsegAvDARgBGDLwsiUbFBBhMNicbh8eatF72GRyRQgZTGZL5Ow9XCiRkwZgID7POY6PSWJHPY4udyeby+CD+QLEYKhbIRKIxOLPRJgBkLdJZHJ5XmFWnFQRlSo1OqNFppdqdFADGSW7qDUbjCV4pjcsnmpYrNa+TbbOX7I73G6Xa7nbR3U6PFwvPZvD5fY4CX5agEgIH7ABCtM5JyhUqCcLzmY5XNRIwATFi8xQqHmXe49F8xMw8uqJJTHLm87C8QGYKjMXjq866PX2SApO83W1JMzO1Lu3mLug+zAK4PcbWRwUdAgxNFWzPBnOpgWpbcQCiYABmStSoebtwFGboAw6KcU2fD0+BABiL4kqIACy3oE95do+DYIDMb4Hu2zhfniP7gpeACsIE0GB84QeyCCNNo75tp+m7fjQADKuErgAbOhJCYVM2HKLYxj4bBRFdiRJCkam4ArgA7DRxB0TC2EgAg6jaDB5KEUeCF5k0YkSSuAAcAk4jW4GjqqzEEYeODHjCHHEAAct8K4AJyTEMwDABOzBYCATiYOg6gABQHBwABW47CE41QgNsLmIFazIAJQhdZpCkG4y6YCA5hiLgUDiHyXD-DqmChOoEDmOoIBZDwVQ8LU5hcLYMBqDAABETT1OYUDyLFzCVc8EiVQA6to5jCAA5F4RkoJVUKRTABy8mIyzKFkuV5M28XCII2RmHISWdBojLKDOUBQCoXihNEOqXGImDRKKrD-jADVbOoYgvu8nZAA))

I am actually not sure there is a way to check this properly.

Comment (nx10) [reply src/util.ts:585]: `Object.keys(ViewColumn).filter(x => isNaN(parseInt(x)))`

Feels a bit hacky but would work.

Comment (nx10) [reply src/util.ts]: fixed now

**Review by renkun-ken — APPROVED** [review]
LGTM

File: src/liveShare/shareSession.ts [hunk 2]
```diff
@@ -70,9 +70,9 @@ export function detachGuest(): void {
 }
 
 export function attachActiveGuest(): void {
-    if (config().get<boolean>('sessionWatcher')) {
+    if (config().get<boolean>('sessionWatcher', false)) {
```
Comment (renkun-ken) [inline src/liveShare/shareSession.ts:73]: The default value of setting `r.sessionWatcher` is `true` already.

File: src/rmarkdown/index.ts [hunk 3]
```diff
@@ -60,17 +60,20 @@ export class RMarkdownCodeLensProvider implements vscode.CodeLensProvider {
...
+            if (!chunk) {
+                continue;
+            }
             const chunkRange = chunk.chunkRange;
             const line = chunk.startLine;
             chunkRanges.push(chunkRange);
 
             // Enable/disable only CodeLens, without affecting chunk background color.
-            if (config().get<boolean>('rmarkdown.enableCodeLens') && (chunk.language === 'r') || isRDocument(document)) {
+            if (config().get<boolean>('rmarkdown.enableCodeLens', false) && (chunk.language === 'r') || isRDocument(document)) {
```
Comment (renkun-ken) [inline src/rmarkdown/index.ts:76]: `r.rmarkdown.enableCodeLens` is by default  `true`.

Comment (renkun-ken) [reply src/rmarkdown/index.ts:76]: I'm curious and wonder if `config().get<T>()` could guarantee that the default value could be returned if somehow missing?

Comment (nx10) [reply src/rmarkdown/index.ts]: I don't think its possible at compile / "type check" time. For correctness I think all occurrences of config().get<T>() should in the future (carefully) be replaced with the defaults. Maybe we could introduce a lint for that.

Comment (nx10) [reply src/rmarkdown/index.ts]: Fixed in 5418748

File: src/liveShare/shareSession.ts (hunk 2 above)
Comment (nx10) [reply src/liveShare/shareSession.ts]: Fixed in 5418748

**Comment by renkun-ken** [comment]
From <https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration>, it seems `config().get()` will at least return the default value if the config entry is found. As long as the extension is properly activated, the config entries should be found anyway. If we always specify a default value, then it might be hard to maintain the settings if we want to change the default, since there might be many `get()` for a certain entry.

Take [vscode-python]() for example, none of the calls to `pythonSettings.get()` (e.g. <https://github.com/microsoft/vscode-python/blob/main/src/client/common/configSettings.ts#L255>) specify the default value since they all access to setting entries defined by the extension. One that specifies the default value is <https://github.com/microsoft/vscode-python/blob/main/src/client/common/configSettings.ts#L182> where it accesses an external entry.

Do you think if it makes sense to follow the pattern that we don't specify the default value if the entry is defined by the extension?

**Comment by nx10** [comment]
> Take [vscode-python](https://github.com/REditorSupport/vscode-R/pull/1209) for example, none of the calls to pythonSettings.get() (e.g. https://github.com/microsoft/vscode-python/blob/a4de2b8a98b1e3ba56f81cc23bd163e6862c059b/src/client/common/configSettings.ts#L255) specify the default value since they all access to setting entries defined by the extension. One that specifies the default value is https://github.com/microsoft/vscode-python/blob/a4de2b8a98b1e3ba56f81cc23bd163e6862c059b/src/client/common/configSettings.ts#L182 where it accesses an external entry.

Interesting, it seems like they opted to do comparisons to avoid `T | undefined`.

> Do you think if it makes sense to follow the pattern that we don't specify the default value if the entry is defined by the extension so that it should automatically follow the definition in package.json?

I was under the impression adding a default value to the call just adds another fallback after there was no default found on any other level. I.e.: `config().get(x) !== undefined ? config().get(x) : defaultValue`

**Comment by renkun-ken** [comment]
It feels quite redundant and burdensome to maintain the default values of each setting not only at where it is defined in `package.json` but also all places that use the setting.

> even in case of typos in the field name

Thought about this case too. And specifying a default when there is a typo seems to hide the problem rather than make it easier to spot.

**Comment by nx10** [comment]
So what is your opinion on how to proceed?

We could also opt for a helper function that either ignores the undefined or checks and errors:

```ts
function config_get<T>(section: string): T {
    return config().get<T>(section) as T;
}

function config_get<T>(section: string): T {
    const x = config().get<T>(section);
    if (x === undefined) {
        throw 'No config entry for ' + section;
        return;
    }
    return x;
}
```

**Comment by renkun-ken** [comment]
>Only using default values when needed is also fine by me, I am just wondering what the most developer friendy way ist.

Me too. Please feel free to use whatever you think makes good sense or common practice.

**Comment by nx10** [comment]
> Please feel free to use whatever you think makes good sense or common practice.

Yes it's probably the best to leave this for now revisit it if it actually presents itself as a problem.

From my side we are good to go with this PR.

**Comment by renkun-ken** [comment]
Thanks!

**Pull request merged**